game-core/src/main/java/games/strategy/triplea/ui/menubar/DebugMenu.java

```diff
@@ -4,13 +4,21 @@
 import games.strategy.triplea.ai.pro.AbstractProAi;
 import games.strategy.triplea.ui.TripleAFrame;
 import java.awt.event.KeyEvent;
+import java.util.Collection;
+import java.util.Map;
 import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Function;
 import javax.swing.JMenu;
+import javax.swing.JMenuItem;
 import org.triplea.swing.SwingAction;
 
-final class DebugMenu extends JMenu {
+public final class DebugMenu extends JMenu {
   private static final long serialVersionUID = -4876915214715298132L;
 
+  private static final Map<String, Function<TripleAFrame, Collection<JMenuItem>>> menuCallbacks =
```

Comment:
The type for this callback seems to be a bit overly generic. It's not clear what the semantic would be behind this type. Perhaps this can be resolved by creating an interface for this, eg: `DebugMenuFactory`

Why is the `TripleAFrame` reference required? Shouldn't any newly created debug menu be able to be added to a generic JMenu? Would it make sense for the debug menu factory to simply produce a menu and let this class deal with the detail of which menu it ought to be added to.

Reply:
I don't understand your issue for the first paragraph.  You want me to create an interface that is an alias?  You mentioned in https://github.com/triplea-game/triplea/pull/7829#discussion_r499175473 about not using interfaces as aliases so I'm confused on why it is ok to use an alias here and not in that previous discussion?

I believe the `TripleAFrame` is required because to create new dialogs and other swing items, a parent frame is required.  So, if the menu item is going to open a new dialog or do something to the UI, it needs access to the `TripleAFrame`.  In one of my debug menu items, it adds an heatmap overlap and it needs to be able to get a reference to the `MapPanel` which is a member of `TripleAFrame`.

Reply:
Presumably it would not be more than just an alias considering `Map<String, Function<TripleAFrame, Collection<JMenuItem>>>` is not a SAM. The interface would give more semantic meaning behind the operation, for example give additional meaning to what the first string parameter represents.

Accessing `TripleAFrame`, or the 'parent frame' in a static way is probably okay and better than passing it around through-out every layer of code (which creates a dependency on Swing and UI throughout the code). As perhaps a better option, there could also be an injection of a UI component into any code that needs to access UI. In this latter suggestion any data to be rendered would be passed into the injected code which would then contain the UI logic for rendering that data.

Reply:
I wasn't able to get an interface to work.  The `= TreeMap<>();` wouldn't compile anymore since `TreeMap` doesn't implement `DebugMenuFactory`.  So, I've added a comment instead to explain what the map is for.

Reply:
Instead of an interface, how about a 1st class object?  The methods on it could then lend it to be more easily understood and otherwise provide a more intuitive API.

Do we really need a TreeMap as well? If the number of elements is just a few, the efficiency gain is seemingly unlikely to be significant.

Reply:
It needs a sorted map so the debug menu items are sorted.  There's only two built in implementations of sortedmap and I don't think `ConcurrentSkipListMap` is right for this situation.

I don't see the need for a 1st class object.  The map is literally a map of a string to a SAM.  It is extremely basic and simple.

Reply:
I don't agree it is simple. Looking at this function:

```Function<TripleAFrame, Collection<JMenuItem>>```

It's not clear how a `TripleAFrame` should transform, or 'map' (functions map one element to another), to a JMenuItem. Why is the frame an input to creating a JMenuItem is a mystery. Beyond that, the type and semantic of the 'string' is not readily understood (stringly typing code-smell).

For example, this could perhaps be instead:
```
List<DebugMenu>
```

Where DebugMenu in an interface with the following API:
```
String getMenuTitle();
JMenu getMenuContents();
```

Reply:
`DebugMenu` would still have a `Function<TripleAFrame, Collection<JMenuItems>>`.  So I think it would be better to just improve the comments.  I understand you are trying to remove the need to use `TripleAFrame` but I don't think that is a possible goal.

I'm fine with changing the `Map<String, Function<TripleAFrame, Collection<JMenuItems>>` to `SortedMap<String, Function<TripleAFrame, Collection<JMenuItems>>` if you are worried that knowing if it is sorted or not is important.